add server 3 version of intro to nomad guide (#6499)
circleci/circleci-docs: jekyll/_cci2/server-3-operator-nomad.adoc

---
version:
- Server v3.x
- Server Admin
---
= Introduction to Nomad Cluster Operation
:page-layout: classic-docs
:page-liquid:
:page-description: Learn how to operate the Nomad Cluster in your CircleCI 2.0 installation.
:icons: font
:toc: macro
:toc-title:

> CircleCI uses https://www.hashicorp.com/blog/nomad-announcement/[Nomad] as the primary job scheduler. This chapter provides a basic introduction to Nomad for understanding how to operate the Nomad Cluster in your CircleCI installation.

Reviewer: A tiny point, but the capitalization of Nomad Cluster/cluster and Job/job is inconsistent.

> CircleCI uses https://www.hashicorp.com/blog/nomad-announcement/[Nomad] as the primary job scheduler. This chapter provides a basic introduction to Nomad for understanding how to operate the Nomad Cluster in your CircleCI installation.

Reviewer: A tiny point, but the capitalization of Nomad Cluster/cluster and Job/job is inconsistent.

toc::[]

== Basic Terminology and Architecture

.Nomad Cluster Management
image::nomad-diagram-v2.png[Diagram of the Nomad cluster]
<<<

> - **Nomad Server:** Nomad servers are the brains of the cluster; they receive and allocate jobs to Nomad clients. In CircleCI, a Nomad server runs on your Services machine as a Docker Container.

Author: ```suggestion
- **Nomad Server:** Nomad Servers are the brains of the Cluster, they receive and allocate jobs to Nomad Clients. In CircleCI, a Nomad Server runs as a service in your Kubernetes cluster.
```

- **Nomad Client:** Nomad clients execute the jobs they are allocated by Nomad servers. Usually a Nomad client runs on a dedicated machine (often a VM) in order to fully take the advantage of machine power. You can have multiple Nomad clients to form a cluster and the Nomad server allocates jobs to the cluster with its scheduling algorithm.

> - **Nomad Jobs:** A Nomad job is a specification, provided by a user, that declares a workload for Nomad. A Nomad job corresponds to an execution of a CircleCI job. If the job uses parallelism, say 10 parallelism, then Nomad will run 10 jobs.

Author: ```suggestion
- **Nomad jobs:** A Nomad job is a specification, provided by a user, that declares a workload for Nomad. A Nomad job corresponds to an execution of a CircleCI job. If the job uses parallelism, say 10 parallelism, then Nomad will run 10 jobs.
```

> - **Build Agent:** Build Agent is a Go program written by CircleCI that executes steps in a job and reports the results. Build Agent is executed as the main process inside a Nomad Job.

Author: ```suggestion
- **Build agent:** Build agent is a Go program written by CircleCI that executes steps in a job and reports the results. Build agent is executed as the main process inside a Nomad job.
```

> == Basic Operations

Author: ```suggestion
== Basic operations
```

The following section is a basic guide to operating a Nomad cluster in your installation.

> The `nomad` CLI is installed in the Nomad pod. It is pre-configured to talk to the Nomad cluster, so it is possible to use the `nomad` command to run the following commands in this section.

Author: ```suggestion
The `nomad` CLI is installed in the Nomad pod. It is pre-configured to talk to the Nomad cluster, so it is possible to use `kubectl` along with the `nomad` command to run the commands in this section.
```

> === Checking the Jobs Status

Author: ```suggestion
=== Checking the jobs status
```

The get a list of statuses for all jobs in your cluster, run:

```shell
kubectl exec -it nomad-server-pod-ID -- nomad status
```

The `Status` is the most important field in the output, with the following status type definitions:

- `running`: Nomad has started executing the job. This typically means your job in CircleCI is started.

- `pending`: There are not enough resources available to execute the job inside the cluster.

- `dead`: Nomad has finished executing the job. The status becomes `dead` regardless of whether the corresponding CircleCI job/build succeeds or fails.

> === Checking the Cluster Status

Author: ```suggestion
=== Checking the cluster status
```

To get a list of your Nomad clients, run:

```shell

> kubectl exec -it nomad-pod-IDm -- nomad node-status

Reviewer 2: ```suggestion
kubectl exec -it nomad-server-pod-ID -- nomad node-status
```

Reviewer 2: Actually I see IDm referenced all over this doc. What is the m?

Author: @ryanwohara It's from here: https://circleci.slack.com/archives/GKZGNMNR5/p1646750514610919?thread_ts=1646750305.583159&cid=GKZGNMNR5
maybe it was a typo?

Reviewer 2: I think it was a typo. I'll make more suggestions so it's easy to resolve.

Author: When copy and paste goes bad! Thank you 🙏

```

NOTE: `nomad node-status` reports both Nomad clients that are currently serving (status `active`) and Nomad clients that were taken out of the cluster (status `down`). Therefore, you need to count the number of `active` Nomad clients to know the current capacity of your cluster.

To get more information about a specific client, run the following from that client:

```shell
kubectl exec -it nomad-server-pod-ID -- nomad node-status -self
```

This will give information such as how many jobs are running on the client and the resource utilization of the client.

> === Checking Logs

Author: ```suggestion
=== Checking logs
```

> As noted in the Nomad Jobs section above, a Nomad Job corresponds to an execution of a CircleCI job. Therefore, Nomad Job logs can sometimes help to understand the status of a CircleCI job if there is a problem. To get logs for a specific job, run:

Author: ```suggestion
As noted in the Nomad jobs section above, a Nomad job corresponds to an execution of a CircleCI job. Therefore, Nomad job logs can sometimes help to understand the status of a CircleCI job if there is a problem. To get logs for a specific job, run:
```

```shell

> kubectl exec -it nomad-pod-IDm -- nomad logs -job -stderr <nomad-job-id>

Reviewer 2: ```suggestion
kubectl exec -it nomad-server-pod-ID -- nomad logs -job -stderr <nomad-job-id>
```

```

NOTE: Be sure to specify the `-stderr` flag as this is where most Build Agent logs appear.

While the `nomad logs -job` command is useful, the command is not always accurate because the `-job` flag uses a random allocation of the specified job. The term `allocation` is a smaller unit in Nomad Job, which is out of scope for this document. To learn more, please see https://www.nomadproject.io/docs/internals/scheduling.html[the official document].

Complete the following steps to get logs from the allocation of the specified job:

. Get the job ID with `nomad status` command.
. Get the allocation ID of the job with `nomad status <job-id>` command.
. Get the logs from the allocation with `nomad logs -stderr <allocation-id>`

> === Shutting Down a Nomad Client

Author: ```suggestion
=== Shutting down a Nomad client
```

When you want to shutdown a Nomad client, you must first set the client to `drain` mode. In `drain` mode, the client will finish any jobs that have already been allocated but will not be allocated any new jobs.

. To drain a client, log in to the client and set the client to drain mode with `node-drain` command as follows:
+
```shell

> kubectl exec -it nomad-pod-IDm -- nomad node-drain -self -enable

Reviewer 2: Ah this section is a bit different from the rest. This is done on the Nomad client where there is no kubectl - they can juts run the nomad commands directly.
```suggestion
nomad node-drain -self -enable
```

Author: Ah right I see, thank you!

```
. Then, make sure the client is in drain mode using the `node-status` command:
+
```shell
nomad node-status -self
```

Alternatively, you can drain a remote node with the following command, substituting the node ID:

```shell
nomad node-drain -enable -yes <node-id>
```

> === Scaling Down the Client Cluster

Author: ```suggestion
=== Scaling down the client cluster
```

> kubectl exec -it nomad-pod-IDm -- nomad node-drain -enable -yes <node-id>

Reviewer 2: ```suggestion
nomad node-drain -enable -yes <node-id>
```
